ts/multicolor_series.ts

```diff
@@ -0,0 +1,729 @@
+            closestPointRange: undefined | number;
+
+        // If the series data or axes haven't changed, don't go through this. Return false to pass
+        // the message on to override methods like in data grouping.
+        if (
+            isCartesian &&
+            !series.isDirty &&
+            !xAxis.isDirty &&
+            !series.yAxis.isDirty && 
+            !force
+        ) {
+            return false;
+        }
+
+        if (processedXData && processedYData) {
+            // Find the closest distance between processed points
+            for (let i = processedXData.length - 1; i >= 0; i--) {
+                distance = processedXData[i] - processedXData[i - 1];
+
+                if (
+                    distance > 0 &&
+                    (typeof closestPointRange === 'undefined' ||
+                    distance < closestPointRange)
+                ) {
+                    closestPointRange = distance;
+
+                    // Unsorted data is not supported by the line tooltip, as well as data grouping and
+                    // navigation in Stock charts (#725) and width calculation of columns (#1900)
+                } else if (distance < 0 && series.requireSorting) {
+                    this.error(15);
+                }
+            }
+
+            // Record the properties
+            series.cropped = cropped; // Type: undefined or true
+            series.cropStart = cropStart;
+            series.processedXData = processedXData;
+            series.processedYData = processedYData;
+
+            if (options.pointRange === null) { // Type: null means auto, as for columns, candlesticks and OHLC
+                series.pointRange = closestPointRange || 1;
+            }
+
+            series.closestPointRange = closestPointRange;
+        }
+
+        return true;
+    }
+
+    public drawTracker (): void {
+        const series = this,
+            options = series.options,
+            trackByArea = options.trackByArea,
+            trackerPath: SeriesColoredSegmentPath[] = 
+                trackByArea ? series.areaPaths : getPath(series.graphPaths),
+            trackerPathLength = isArray(trackerPath) && trackerPath.length,
+            chart = series.chart,
+            pointer = chart.pointer,
+            renderer = chart.renderer,
+            snap = chart.options.tooltip?.snap ?? 0,
+            tracker = series.tracker,
+            cursor = options.cursor,
+            css = cursor && { cursor },
+            singlePoints = series.singlePoints,
+            trackerFill = 'rgba(192,192,192,0.002';
+
+        let singlePoint,
+            i;
+
+        const onMouseOver = function (): void {
+            if (chart.hoverSeries !== series) {
+                series.onMouseOver();
+            }
+        };
+        
+        // Extend end points. A better way would be to use round linecaps,
+        // but those are not clickable in VML.
+        if (trackerPathLength && !trackByArea) {
```

Comment:
I'd keep it inside the scope of the `drawTracker` method since the logic isn't reused somewhere else. But hmm, of course, we can do that, but maybe we should extract the below _Handle single points_ section as well. What do you think?

Reply:
Yes, we can extract Handle single points section as well. It is a good pattern to keep methods relatively small and logically consistent. Testing and maintaining functions with less logic is also simpler.

Reply:
Changed, leaving this comment as unresolved to make sure you'll check it 😀